crates/extensions/c8y_mapper_ext/src/actor.rs

```diff
@@ -95,25 +120,46 @@ impl Actor for C8yMapperActor {
             .send(SyncStart::new(SYNC_WINDOW, ()))
             .await?;
 
-        while let Some(event) = self.messages.recv().await {
-            match event {
-                C8yMapperInput::MqttMessage(message) => {
-                    self.process_mqtt_message(message).await?;
-                }
-                C8yMapperInput::FsWatchEvent(event) => {
-                    self.process_file_watch_event(event).await?;
-                }
-                C8yMapperInput::SyncComplete(_) => {
-                    self.process_sync_timeout().await?;
-                }
-                C8yMapperInput::IdUploadResult((cmd_id, result)) => {
-                    self.process_upload_result(cmd_id, result).await?;
-                }
-                C8yMapperInput::IdDownloadResult((cmd_id, result)) => {
-                    self.process_download_result(cmd_id, result).await?;
+        let mut messages = self.messages;
+
+        let ops_dir = Arc::clone(&self.converter.config.ops_dir);
+        let worker = Arc::new(C8yMapperWorker {
+            converter: Mutex::new(self.converter),
+            mqtt_publisher: self.mqtt_publisher,
+            ops_dir,
+        });
+
+        while let Some(event) = messages.recv().await {
+            let worker = Arc::clone(&worker);
+
+            tokio::spawn(async move {
```

Comment:
I had a different model in mind where we have independent workers for each operation handler with minimal coordination with the top level mapper actor/worker, eliminating the need to share and lock the top level converter. This mapper actor would still coordinate the MQTT messages and convert them to typed operation messages and dispatch that to those respective operation workers. Once the worker gets an operation message, they should be able to freely exchange messages between the other actors like upload/download actors. Extending the usage of the `RequestEnvelope` beyond the HTTP actors would be key to enabling this.

Reply:
> I had a different model in mind where we have independent workers (even actors, may be) for each operation handler with minimal coordination with the top level mapper actor/worker, eliminating the need to share and lock the top level converter.

Yes, the idea would be to decouple as much as possible from `CumulocityConverter` wrt. the operation handling, so it can be run separately as a new task, or maybe as a new actor, without locking the rest of the converter. Right now, basically all the functionality is implemented in `CumulocityConverter` in `&mut self` methods, so it's impossible to run stuff concurrently without locking.

Ideally the each operation handler should be as independent as possible, and we should be able to easily reason about each in isolation, without the control flow jumping between `actor.rs`, `converter.rs`, and e.g. `log_upload.rs`.

Right now they're just tokio tasks, because my current goal is to just move the boundary on which we block, but the handlers could definitely be extracted out of the main actor.

> This mapper actor would still coordinate the MQTT messages and convert them to typed operation messages and dispatch that to those respective operation workers. Once the worker gets an operation message, they should be able to freely exchange messages between the other actors like upload/download actors. Extending the usage of the RequestEnvelope beyond the HTTP actors would be key to enabling this.

That's right, the actor which owns the MQTT message box should route these operation state change message to the operation handlers, which should then be able to do do blocking stuff, like wait for the upload, without synchronizing with the main actor. But I'd say that we overuse messages a bit. It's fine to have actors that manage shared mutable state, but for downloading/uploading, what shared state is there that needs to be managed between different downloads/uploads? Why shouldn't we use the downloader/uploader directly instead of going via the actor?

And I believe that sometimes we use the actor only so that we can mock it in tests. That instead of having an interface that we can mock, we effectively mock the interface by simulating messages being sent in the message boxes. And I think this way of testing is a bit tedious, because you need to handle and ignore unrelated messages, and sometimes when messages are not being send by actor under test, you need to do timeouts, and so on. I think there's room for improvement there.

But that's a discussion for another time, and I expect we'll discuss it more in the future. For the purposes of this PR, the important part is allowing handlers to await either tasks they themselves spawn, or requests sent to other actors via `RequestEnvelope`, one of these two.

Reply:
> Why shouldn't we use the downloader/uploader directly instead of going via the actor?

One additional reason, other than message visibility and testability, is to control the bandwidth usage as well. We don't want to let too many uploads/downloads happening parallelly, to reduce the load on the network and the disk. Routing those via a centralised downloader/uploader will let us better control the maximum parallelism that can be allowed.